### DISCUSSION TITLE
ESP32 and servo SG90

### DISCUSSION THREAD

**Discussion by Mi-Dash** [arendst/Tasmota#20512] [Q&A]
Hi!
I'm trying to control an SG90 servo with ESP32-DevKit without success.
I am following the "Servo Motor Control HOWTO #10443".
I checked it on the ESP8266, all works fine but I can’t repeat it on the ESP32.
Firmware downloaded from https://ota.tasmota.com/tasmota32/release/tasmota32.bin

Here's a slightly truncated console output:

_"Info1":{"Module":"ESP32-DevKit","Version":"13.3.0(tasmota32)"_

_CMD: SetOption80
RSL: RESULT = {"SetOption80":"ON"}
CMD: ShutterMode
RSL: RESULT = {"ShutterMode":5}
CMD: PWMfrequency
RSL: RESULT = {"PWMFrequency":200}
CMD:SetOption15
RSL: RESULT = {"SetOption15":"OFF"}_

Everything is fine but

_CMD: ShutterPwmRange1 100, 500
RSL: RESULT = {"Command":"Error"}_

And after _restart 1_

_CMD: Shuttermode
RSL: RESULT = {"ShutterMode":0}_

What i'm doing wrong?

**Comment by Mi-Dash** [comment]
I tried checking different firmware versions. In my case version 10.1.0 (and maybe 10.0.0) works fine. I tried version 9.1.0 and below, but found some bugs in earlier versions.

Starting from version 11.0.0 the _PwmFrequency_ parameter does not work. The PWM frequency does not change and always remains about 1000 (1ms), which leads to incorrect pulse generation and incorrect servo control. The latest version 13.3.0 does not understand the _ShutterPwmRange_ parameter and resets _Shuttermode_ to 0.

I haven't checked absolutely all versions. I don’t know starting from which version the _ShutterPwmRange_ parameter does not work. 

Now my choice is 10.1.0 and my configuration is {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,224,224,1,1,1312,1,416,1,1,1,258,260,0,261,259,1,0,1,1,230,0,0,0,0,640,608,1,1,1,0,0,1],"FLAG":0,"BASE":1}

12.2.0 and 13.3.0 on ESP8266 hasn't problem

**Reply by stefanbode** [reply]
Hi, I will take a look. The PWM handler are different on ESP8266 and ESP32. The complete story works a bit different. I made some fixes a few month ago bit not sure if only on the esp8266 side. Have to set up a testbench. Thank god its weekend

**Reply by stefanbode** [reply]
I see that after a reboot the shuttermode is 1. But if you set it to shuttermode 5 again then it stays there. Also now after a reboot the PWMrange is working
![image](https://github.com/arendst/Tasmota/assets/16225290/40275efe-3c52-45bf-8b70-f4dbc0e4310c)

**Reply by stefanbode** [reply]
the esp32 has more complex options to store and control. it uses the file drvset027 for this purpose. Looks like the shuttermode 5 in the first round does not make it into the file.

**Reply by stefanbode** [reply]
I have submitted a change that ensure proper init procedure. With the change your problems during setup should be away. 
Pull request #20548

**Reply by Jason2866** [reply]
PR is merged

**Comment by Jason2866** [comment]
@stefanbode Can you have a look?

**Comment by Mi-Dash** [comment]
Thank you!
I need some time to get back to the test. I will definitely check and report the result.

**Comment by Mi-Dash** [comment]
Where can I get the correct new firmware? 13.3.0.3 with timestamp 20240121 23:19 from development also resets the _Shuttermode_ to 0 after restart.
Sorry, I haven't learned much yet

**Reply by stefanbode** [reply]
Can you please set loglevel 4 and send some log files. As soon as you set the shuttermode all gets saved into flash. Unlikely that it went back to 0. have to see more

**Comment by Mi-Dash** [comment]
That is right?
![2024-01-23_08-41-27](https://github.com/arendst/Tasmota/assets/85215254/1babafa9-4bbc-422a-982f-2bf855544f40)
![2024-01-23_08-42-02](https://github.com/arendst/Tasmota/assets/85215254/f2a37543-951e-4e13-a679-742334775149)

**Reply by stefanbode** [reply]
Yes this helps and is really strange. Setoption80 1 is set I assume. Normally you should see an every command that it writes into the file system. Nothing....
Can you check the answer on "so80" command?

**Comment by Mi-Dash** [comment]
![2024-01-23_21-13-26](https://github.com/arendst/Tasmota/assets/85215254/7f4f6bed-4a67-41d2-8e13-5ad4ef922c27)

**Comment by Mi-Dash** [comment]
![2024-01-23_21-26-27](https://github.com/arendst/Tasmota/assets/85215254/26041f32-35e4-40d5-9a92-595b7f94b6fe)

**Comment by Mi-Dash** [comment]
I have previously entered all the parameters from the https://github.com/arendst/Tasmota/discussions/10443". Everything is still after reboot and even after changing the firmware. But not the _Shuttermode_.

**Reply by stefanbode** [reply]
Agree. Very strange. I will check with some other developers what can be the root cause. The esp32 saves its shutter data to the file system. I assume this is always available. But currently it looks like I am wrong. In the Menu "CONSOLE" there should be a sub-menu file system. Is you have teams or discord it might be best is we do a sharing session. As said: very uncommon behavior. My Timezone is Germany. What's yours ?

**Reply by Mi-Dash** [reply]
I'll be glad to connect. See email.

**Comment by Mi-Dash** [comment]
![2024-01-24_14-02-06](https://github.com/arendst/Tasmota/assets/85215254/80f31d32-484e-4069-8d54-e72129971741)
